### ISSUE TITLE
Drop placeholder "performance" issue when different height to items

### ISSUE THREAD

**Issue by mataslib** [isaacHagoel/svelte-dnd-action#310]
When the placeholder is different than the item, specifically its height, bad behavior occurs, dnd becomes unresponsive (or calculations break).

Repl: https://svelte.dev/repl/74a3b7658a8b40ba99a660ea9eb33054?version=3.41.0

**Comment by isaacHagoel** [comment]
i need a bit more details about the behaviour. the dragged element copies it the styles of the shadow element (top level, ignoring children styles). 
Do you want the dragged element to morph to the size of the small placeholder? that should be achievable

**Comment by mataslib** [comment]
I would like this design to work correctly. (I updated css of repl to match this gif). 

I![2021-08-04-08-45-06 (1)](https://user-images.githubusercontent.com/5154014/128144891-9ee4f03b-3fc6-48f1-9453-35e1303fc839.gif)

**Comment by mataslib** [comment]
Got back to this today and did some debugging, it seems that custom placeholder is not the issue but it's caused by different item heights.

No problem dragging big items ("responsive"), but becomes "unresponsive" when dragging small ones.

![2021-09-07-13-00-58](https://user-images.githubusercontent.com/5154014/132334116-709c33d1-9196-445b-948a-c0892d9d5f06.gif)


Repl: https://svelte.dev/repl/97224142fc5641c4849c8ab55ad473bd?version=3.24.1

**Comment by mataslib** [comment]
@isaacHagoel 

Problems with current `wouldBeIndex` implementation:

A) Different dragging behavior on way back
- When you go top-bottom shadow moves when center of dragged ENTERS i-th child rect
- On way back bottom-top shadow moves when center of dragged LEAVES i-th child - to be precise, when dragged is over cached shadow rect (not visible), which was cached on way top-bottom

REPL: https://svelte.dev/repl/97224142fc5641c4849c8ab55ad473bd?version=3.24.1
GIF: 
![2021-09-08-13-23-25](https://user-images.githubusercontent.com/5154014/132501008-cda55cd6-c838-4eed-a868-bb5483821709.gif)

B) Dnd becomes "unresponsive" on fast movement back.
REPL: https://svelte.dev/repl/97224142fc5641c4849c8ab55ad473bd?version=3.24.1
GIF: ![2021-09-08-13-27-24](https://user-images.githubusercontent.com/5154014/132501445-5b7afff1-7205-4daf-bf0d-aa74e1f66833.gif)

**Comment by mataslib** [comment]
I created draft PR that solves this ONLY for vertical dnd zone. I think UX in draft is better than actual, same behavior for dragging both ways and always "responsive". However it's based on drag direction and I have no idea how to make it work for vertical and horziontal (and possibly grid) lists without adding config option forcing user to specify.

**Comment by isaacHagoel** [comment]
the caching mechanism was created on purpose to solve issues with items jumping around.
getting locked on fast drag back is definitely a bug that i want to fix.
can your solution be enhanced to be container direction independent?

**Comment by isaacHagoel** [comment]
@mataslib ^

**Comment by mataslib** [comment]
@isaacHagoel Hello isaac.

> can your solution be enhanced to be container direction independent?

I don't know, I don't think so, I didn't figure out any idea yet. 

I think that https://github.com/isaacHagoel/svelte-dnd-action/issues/312 may be duplicate of this issue.

Unfortunately I'm finishing project and can not spend any more time on this as of now. I didn't realize how complex this is and how many strategies are there. It seems like other libs also have problems supporting grids and rather supports Vertical/Horizontal lists and custom user strategies. I will spend some time on container autoscroll support now as it blocks shipping the project.

**Comment by isaacHagoel** [comment]
> B) Dnd becomes "unresponsive" on fast movement back.
> REPL: https://svelte.dev/repl/97224142fc5641c4849c8ab55ad473bd?version=3.24.1

Looking at the REPL and gif again (wanted to fix it) - looks like it is working as intended to me. In the gif the item is not dragged back high enough to make the large item move. I don't think it is related to the speed of dragging

**Comment by mataslib** [comment]
> looks like it is working as intended to me. In the gif the item is not dragged back high enough to make the large item move.

In my opinion it is 130px over first large item "Am", that seems enough to me to make it moved? I think the correct behavior is that shadow is never allowed to be in further distance than 1 idx away from bellow element. It's distance of two in gif and I am even able to make it bigger distance as you can in this screenshot.
![image](https://user-images.githubusercontent.com/5154014/133029822-91a859bc-7cea-41c8-9bf7-8d27561736cd.png)

Not mentioning difference in behavior on switching direction of dragging. How can user know where to drag after a while of dragging - is it start of element or end?

**Comment by isaacHagoel** [comment]
This was the original behaviour but that would cause the large item to
bounce back and forth as the small item was dragged above it because
depending on where it was the small item would decide it needs to be index
0 or index 1. with the new behaviour the item will move the first time and
then cache the position of its shadow element, any subsequent move in any
direction would depend on that (== it would trigger movement only if it is
above where it would be if dropped).
a perfect solution would know the location of the shadow item in advance
without having to "see" it the first time it moves. that would conflict
with one of the library's design goals == not assuming the items' layout.
users have not complained about this behaviour as far as I know (and the
lib is used in production)

**Comment by mataslib** [comment]
I get your point, all hat I can say is that there are many things in production which doesn't make them correct even if nobody complained yet. (Actually I'm the first one, our designer second one, hopefuly there will be others.)

Back to the problem. I don't get one thing, In the last image, shadows were cached on the way down. How is it possible, that above element is hovering over item idx 0, but shadow of item idx 4 is rendered. I would expect shadow at idx 0 or 1.

**Comment by isaacHagoel** [comment]
you are right. in the second example the index should become 2 (between the two large items). this is definitely something i can look into and fix

**Comment by kkarpeev** [comment]
Hi! We have the same problem. Can you, please, tell what's the status?
I see that @mataslib has created a pull request to fix this. Is something wrong with it?

**Comment by mataslib** [comment]
@kkarpeev 

Well, there are multiple problems with it. Main one is that it's currently solving only Vertical dropzone and it's hardcoded. It would have to be further updated to at least allow picking between default one and my algo, optionally adding support for horizontal dropzone too. PR algo also forces user to specify whether dropzone is vertical or horizontal, and prob wouldn't work with grid dropzone (vertical + horizontal).  Also I have updated version of this pr in my project (some bugs fixed), since we don't use grids, this PR works nicely for our usecase. Unfortunately i don't have resources to spend more time on this at the moment or in near future. It would be nice to allow user to pass his own strategy (or picking library's one) as there are many strategies of collision detection to pick from (nice blogpost that inspired my pr is https://docs.dndkit.com/api-documentation/context-provider/collision-detection-algorithms), fix or change current solution. In general react dnd libraries also specifies whether dropzone is horizontal or vertical for optimal experience.

**Comment by kkarpeev** [comment]
@mataslib Oh, I see. Thank you!

**Comment by benkuper** [comment]
Reviving an old thread, but I've got the same problem when using morph. Also, when the dragged element is morphing back and forth, the dragged element is not recentering relative to mouse and can be way off the actual mouse position.
Those are 2 problems, but I believe that at least being able to choose mouse position as collision system would help a lot on stability. What do you think ? I can make another issue if you prefer.

Swap thresholds in SortableJS was what made me feel good about this concept : https://sortablejs.github.io/Sortable/
